Этап 2: Мапперы, Бд, Модели и структура

oblanox opened the pull request:
Частично немного забежали в Этап 3

oblanox committed:
Merge pull request #21 from IP-Kuzmin/develop

Дополнены тесты - требуется БД

oblanox committed:
Merge pull request #23 from IP-Kuzmin/develop

fix: поправлены элементы DTO

oblanox committed:
feat: добавлены мапперы
feat: добавлены прототипы сервисов
feat: добавлена авторизация пользователя
feat: добавлен пароль и секьюрити пароля
feat: добавлены PostgreSQL поддержка
feat: добавлен liquibase
feat: частично добавлен logger
fix: настроен lombok
feat: добавлены тесты по Авторизации и Профилю Пользователей

oblanox committed:
fix: расширенный лог фронта перемещен в filter package

oblanox committed:
fix: расширенный лог фронта перемещен в filter package

AleksandrZhukovJava commented on src/main/java/ru/skypro/homework/config/WebSecurityConfig.java:
```diff
@@ -59,4 +42,8 @@ public PasswordEncoder passwordEncoder() {
         return new BCryptPasswordEncoder();
     }
 
+    @Bean
+    public UserDetailsService userDetailsService(ru.skypro.homework.repository.UserRepository userRepository) {
+        return new AuthUserDetailsService(userRepository);
+    }
```
Сделайте сам класс компонентом, не создавайте вручную

AleksandrZhukovJava commented on src/main/java/ru/skypro/homework/dto/Ad.java:
```diff
@@ -7,13 +7,19 @@
-    @Schema(description = "ID объявления", example = "1")
-    private Integer id;
+
+    @Schema(description = "ID объявления (primary key)")
```
primary key это наша бэковая структура, фронту это не интересно

AleksandrZhukovJava commented on src/main/java/ru/skypro/homework/exceptions/ExceptionsMessage.java:
```diff
@@ -0,0 +1,22 @@
+
+    private final String message;
+
+    ExceptionsMessage(String message) {
```
private иначе всю концепцию ломаем

AleksandrZhukovJava replied:
Upd: Внезапно вспомнил что он по умолчанию private - отменяется

AleksandrZhukovJava commented on src/main/java/ru/skypro/homework/filter/RequestLoggingFilter.java:
```diff
@@ -0,0 +1,32 @@
+@Component
+@RequiredArgsConstructor
+@Slf4j
+public class RequestLoggingFilter extends OncePerRequestFilter {
```
уже был вроде такой

oblanox replied:
Перенесен из component, подсунул в фильтры. Без него отлаживать неудобно - потом потру.

AleksandrZhukovJava commented on src/main/java/ru/skypro/homework/service/auth/AuthUserDetailsService.java:
```diff
@@ -0,0 +1,28 @@
+import ru.skypro.homework.models.UserModel;
+import ru.skypro.homework.repository.UserRepository;
+
+public class AuthUserDetailsService implements UserDetailsService {
```
@Component

AleksandrZhukovJava commented on src/main/java/ru/skypro/homework/service/impl/AdServiceImpl.java:
```diff
@@ -1,57 +1,87 @@
     @Override
     public void deleteAd(Integer id) {
-        // Пустой метод для тестов
+        adRepository.deleteById(id);
```
Пользователю надо сообщить что мы не удалили не существующую запись и сообщить что удалили если смогли, он мог например ввести не тот id и будет в полной уверенности что всё прошло отлично

oblanox replied:
Да, я обычно не оставляю "вникуда" репу, она не мокается. Это в рефакторе все будет. Но репа вникуда - это у меня лично - ошибка, да)

AleksandrZhukovJava commented on src/main/java/ru/skypro/homework/service/impl/AdServiceImpl.java:
```diff
@@ -1,57 +1,87 @@
     @Override
     public Ads getUserAdsCount() {
-        return new Ads(0, Collections.emptyList());
+        UserModel author = userRepository.findById(1L).orElseThrow(); // заглушка
```
Почему 1L? и нужен кастомный exception
Если заглушка, то отглушайте, уже пора =)

oblanox replied:
Мы бедные маленькие ниндзя-черепашки, отягощенные несправедливыми кредитами и бесцельными трудами, времени не дают - телефонят заказчики, а телефон не отключить... Сегодня всё будет....

AleksandrZhukovJava commented on src/main/java/ru/skypro/homework/service/impl/AuthServiceImpl.java:
```diff
@@ -1,29 +1,59 @@
+        UserModel user = userRepository.findByEmail(login.getUsername())
+                .orElseThrow(() -> {
+                    log.error("Пользователь {} не найден", login.getUsername());
+                    System.out.println(new BCryptPasswordEncoder().encode("password"));
```
енкодер сделать бином и убрать в секьюрити конфиг иначе создаёте каждый раз

AleksandrZhukovJava commented on src/main/java/ru/skypro/homework/service/impl/AuthServiceImpl.java:
```diff
@@ -1,29 +1,59 @@
+
+        if (!passwordEncoder.matches(login.getPassword(), user.getPassword())) {
+            log.warn("Неверный пароль для пользователя {}", login.getUsername());
+            throw new ResponseStatusException(HttpStatus.UNAUTHORIZED, ExceptionsMessage.WRONG_PASSWORD.getMessage());
```
Это не бизнесовый Exception, отразите в названии суть, тут и далее
Пере использование это хорошо, но тут это плохо масштабируется, малейшее изменение сигнатуры и этого исключения уже не хватит и тогда будет не понятн ов чем вообще был смысл.
Саму идею с сообщениями в Enum можно оставить (хотя оверхэд чуть чуть) но вот ошибки везде нужны кастомные и разные

AleksandrZhukovJava commented on src/main/java/ru/skypro/homework/service/impl/CommentServiceImpl.java:
```diff
@@ -1,35 +1,72 @@
+    public Comment addComment(Integer adId, CreateOrUpdateComment dto) {
+        AdModel ad = adRepository.findById(adId)
+                .orElseThrow(() -> new ResponseStatusException(NOT_FOUND, "Объявление не найдено"));
+        UserModel author = userRepository.findById(1L)
```
Почему 1L

oblanox replied:
Недописано, я прописал только половину UserService...уже правда полностью (пока ждал домашку), но AdService и CommentService - пока Болванчики, завтра уже пропишу и сдам.

AleksandrZhukovJava replied:
Уже понял когда посмотрел какая неделя)

AleksandrZhukovJava commented on src/main/java/ru/skypro/homework/service/impl/CommentServiceImpl.java:
```diff
@@ -1,35 +1,72 @@
-        return new Comment("БаБаБла");
+    public Comment addComment(Integer adId, CreateOrUpdateComment dto) {
+        AdModel ad = adRepository.findById(adId)
+                .orElseThrow(() -> new ResponseStatusException(NOT_FOUND, "Объявление не найдено"));
```
Например тут не передать какое конкретно объявление не найдено. А без этого как потом искать ошибку по логам? Даже если её занести внутрь сообщения потом это будет тяжело обрабатывать (не будет поля) и локализировать

oblanox replied:
То же самое, это прототип - завтра отрефакторю.

AleksandrZhukovJava commented on src/main/java/ru/skypro/homework/service/impl/UserServiceImpl.java:
```diff
@@ -1,43 +1,99 @@
+        }
+
+        log.info("Пользователь {} успешно сменил пароль", user.getEmail());
+        log.debug("Новый пароль: {}", passwords.getNewPassword());
```
никогда не логируйте такие чувствительные данные, даже дебагом

AleksandrZhukovJava commented on src/main/java/ru/skypro/homework/service/impl/UserServiceImpl.java:
```diff
@@ -1,43 +1,99 @@
-                Role.USER,
-                "https://example.com/image.jpg"
-        );
+        System.out.println("jjjj");
```
убрать

AleksandrZhukovJava commented on src/main/java/ru/skypro/homework/service/impl/UserServiceImpl.java:
```diff
@@ -1,43 +1,99 @@
+        log.debug("Обновлённые поля: firstName={}, lastName={}, phone={}",
+                dto.getFirstName(), dto.getLastName(), dto.getPhone());
+
+        return userMapper.toDto(userRepository.save(user));
```
Вот тут сохранение очень легко пропустить, его не то что надо вынести еще и пробелами отделить от остального кода

oblanox committed:
fix: исправлены модели UserModel, AdModel, CommentModel с аннотациями Hibernate и Swagger
fix: улучшены DTO классы согласно OpenAPI-спецификации
fix: изменены контроллеры UserController, AdController, CommentController, AuthController с аннотациями OpenAPI
feat: добавлены сервисы UserServiceImpl, AdServiceImpl, CommentServiceImpl, AuthServiceImpl с базовой логикой и логированием
fix: исправлен маппинг между сущностями и DTO через UserMapper, AdMapper, CommentMapper
feat: добавлена авторизация и регистрация пользователя с использованием Spring Security
feat: реализована логика смены пароля пользователем и обновления профиля
feat: реализовано добавление, обновление, получение и удаление объявлений с загрузкой изображений
feat: реализовано добавление, обновление, получение и удаление комментариев к объявлениям
feat: реализовано получение объявлений и комментариев авторизованного пользователя
feat: добавлены валидаторы и сообщения об ошибках для всех DTO согласно OpenAPI
feat: добавлены глобальные исключения и собственные Exception-классы
fix: исправлен маппинг ExtendedAd с учётом конструктора с полным набором полей
fix: обновлён AdServiceImpl под реальные изображения и работу с репозиторием
fix: реализована загрузка аватарки пользователя и сохранение в файловую систему
fix: исправлена логика loginUser и registerUser для обработки ошибок и соответствия статусам
fix: корректно возвращаются 400/401/403/404 коды ошибок по OpenAPI спецификации
fix: добавлены тесты для всех контроллеров: AuthController, UserController, AdController, CommentController
feat: добавлены custom Exceptions
